Scripts/Runtime/Entities/TaskDriver/TaskSet/TaskData/DataStream/Cancellation/CancelProgressFlowNode.cs

```diff
@@ -0,0 +1,194 @@
+        private readonly ActiveLookupData<EntityProxyInstanceID> m_ProgressLookupData;
+        private readonly ActiveLookupData<EntityProxyInstanceID> m_ParentProgressLookupData;
+        private readonly PendingData<EntityProxyInstanceID> m_CancelCompleteData;
+        private readonly uint m_CancelCompleteActiveID;
+
+        private NativeArray<JobHandle> m_Dependencies;
+
+
+        public CancelProgressFlowNode(ITaskSetOwner taskSetOwner, CancelProgressFlowNode parent)
+        {
+            m_TaskSetOwner = taskSetOwner;
+            m_Parent = parent;
+
+            m_ProgressLookupData = m_TaskSetOwner.TaskSet.CancelProgressDataStream.ActiveLookupData;
+            m_CancelCompleteData = m_TaskSetOwner.TaskSet.CancelCompleteDataStream.PendingData;
+            m_CancelCompleteActiveID = m_TaskSetOwner.TaskSet.CancelCompleteDataStream.GetActiveID();
+
+
+            m_Dependencies = new NativeArray<JobHandle>(4, Allocator.Persistent);
+
+            if (m_Parent != null)
+            {
+                m_ParentProgressLookupData = m_Parent.m_TaskSetOwner.TaskSet.CancelProgressDataStream.ActiveLookupData;
+            }
+        }
+
+        protected override void DisposeSelf()
+        {
+            m_Dependencies.Dispose();
+            base.DisposeSelf();
+        }
+
+        public override string ToString()
+        {
+            return $"{m_TaskSetOwner} - With CancelRequestData of {m_TaskSetOwner.TaskSet.CancelRequestsDataStream.GetActiveID()} and CancelProgressData of {m_TaskSetOwner.TaskSet.CancelProgressDataStream.ActiveLookupData.ID} and CancelCompleteData of {m_TaskSetOwner.TaskSet.CancelCompleteDataStream.GetActiveID()}";
+        }
+
+        private JobHandle ScheduleCheckCancelProgressJob(JobHandle dependsOn)
+        {
+            m_Dependencies[0] = dependsOn;
+            m_Dependencies[1] = m_ProgressLookupData.AcquireAsync(AccessType.ExclusiveWrite);
+            m_Dependencies[2] = m_CancelCompleteData.AcquireAsync(AccessType.SharedWrite);
+            m_Dependencies[3] = (m_Parent != null)
+                ? m_ParentProgressLookupData.AcquireAsync(AccessType.ExclusiveWrite)
+                : default;
+
+            dependsOn = JobHandle.CombineDependencies(m_Dependencies);
+
+            CheckCancelProgressJob checkCancelProgressJob = new CheckCancelProgressJob(m_ProgressLookupData.Lookup,
+                                                                                       m_CancelCompleteData.PendingWriter,
+                                                                                       m_CancelCompleteActiveID,
+                                                                                       m_ProgressLookupData.TaskSetOwner.ID,
+                                                                                       m_Parent != null
+                                                                                           ? m_ParentProgressLookupData.Lookup
```

Comment:
I've referred to this in #136. 

I'm not sure yet on it. I agree we want to make it more safe though. My issue is that there are times I need the data inside because I want to operate on it. Which in that case it makes perfect sense to not let you touch it unless you have the access. 

But for other times, I'm building structures where I just need reference to the internal data and I know i'll have access later when I want to use it. 

But it's a good point and something I'll look at in more depth with #136

Reply:
Yea I'd question the pattern being used where you are gathering references but don't have access rights. You're circumventing safety at that point. Definitely convenient but maybe indicative of a pattern that's going to cause you issues in the future.

Reply:
That's a good point. Something to make sure of when cleaning up the internals.